system-contracts/contracts/EvmInterpreterLoop.template.yul

```diff
@@ -459,12 +459,7 @@ for { } true { } {
             revertWithGas(evmGasLeft)
         }
 
-        for { let i := 0 } lt(i, len) { i := add(i, 1) } {
-            mstore8(
-                add(dst, i),
-                shr(248, mload(add(offset, i)))
-            )
-        }
+        mcopy(dst, offset, len)
```

Comment:
@jrchatruc Is there a chance that dst and offset pointers can overlap? If that can never happen, we could instead of calling mcopy (essentially optimized implementation of memmove) we could implement memcpy that is efficient that this.

Reply:
Yes, `dst` and `offset` can overlap, from `MCOPY` in https://www.evm.codes/ :

> Copying takes place as if an intermediate buffer was used, allowing the destination and source to overlap

Reply:
Since this is change for `OP_CODECOPY`, my question was whether `dst` and `offset` can overlap in for opcode? Please note that mcopy function handles overlap case.

Reply:
No they can't, since `offset` is from the bytecode memory space, while `dst` is from the heap memory space.
A check is being performed to make sure they are inside their respective memory space.

This is the check:

```
        // Check bytecode overflow
        if gt(add(offset, len), sub(MEM_OFFSET(), 1)) {
            revertWithGas(evmGasLeft)
        }
```

Reply:
In that case, I have implemented memcpy function that is faster than mcopy, since mcopy handles overlap  and here, we don't need to worry about.